### PR TITLE
Q_Numtraits

### PR DESCRIPTION
**Pull request by 3pmile**
**Description**

This PR implements...
- [x] is_zero for Q, MatQ, PolyOverQ
- [x] is_one for Q, PolyOverQ
- [x] is_square and is_identity for MatQ

as a new https://github.com/qfall/math/pull/271 .

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/rational/mat_q/properties.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+//! This module includes functionality about properties of [`MatQ`] instances.
+
+use super::MatQ;
+use flint_sys::fmpq_mat::{fmpq_mat_is_one, fmpq_mat_is_square, fmpq_mat_is_zero};
+
+impl MatQ {
+    /// Checks if a [`MatQ`] is the identity matrix.
+    ///
+    /// Returns true if every diagonal entry is `1` and all other entries are
+    /// `0`.
```
Comment (MoogSven) [inline src/rational/mat_q/properties.rs:18]: Explicitly mention that non quadratic matrices are allowed.

File: src/rational/mat_q/properties.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    ///
```
Comment (MoogSven) [inline src/rational/mat_q/properties.rs:19]: `# Examples` missing here and at the next two functions

File: src/rational/mat_q/properties.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        1 == unsafe { fmpq_mat_is_zero(&self.matrix) }
+    }
+}
+
+#[cfg(test)]
+mod test_is_identity {
+    use super::MatQ;
+    use std::str::FromStr;
+
+    /// ensure that is_identity returns `true` for identity matrices
```
Comment (MoogSven) [inline src/rational/mat_q/properties.rs:67]:
```suggestion
    /// Ensure that is_identity returns `true` for identity matrices.
```
Start with upper case and end with a dot as discussed today. Also applies to the next comments.

File: src/rational/mat_q/properties.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    #[test]
+    fn identity_detection() {
+        let ident = MatQ::from_str("[[1, 0],[0, 1]]").unwrap();
+
+        assert!(ident.is_identity());
+    }
+
```
Comment (MoogSven) [inline src/rational/mat_q/properties.rs:74]: Add a test for a non square identity.

File: src/rational/mat_q/properties.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    /// ensure that is_square returns `false` for non-square matrices
+    #[test]
+    fn sqaure_rejection() {
+        let small = MatQ::from_str("[[0, 5/6, 4],[2/7, 0, 1]]").unwrap();
+        let large = MatQ::from_str("[[9, 0],[127/71, 0],[0, 0]]").unwrap();
+
+        assert!(!small.is_square());
+        assert!(!large.is_square());
```
Comment (MoogSven) [inline src/rational/mat_q/properties.rs:132]: Variable names do not really fit here. Maybe call them `mat_2x3` and `mat_3x2` instead
```suggestion
        let mat_2x3 = MatQ::from_str("[[0, 5/6, 4],[2/7, 0, 1]]").unwrap();
        let mat_3x2 = MatQ::from_str("[[9, 0],[127/71, 0],[0, 0]]").unwrap();

        assert!(!mat_2x3.is_square());
        assert!(!mat_3x2.is_square());
```

File: src/rational/poly_over_q/properties.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+//! This module includes functionality about properties of [`PolyOverQ`] instances.
+
+use flint_sys::fmpq_poly::{fmpq_poly_degree, fmpq_poly_is_one};
+
+use super::PolyOverQ;
+
+impl PolyOverQ {
+    /// Checks if a [`PolyOverQ`] is the constant polynomial with coefficient `1`.
+    ///
+    /// Returns true if the first coefficient is `1` and is the only coefficient
```
Comment (MoogSven) [inline src/rational/poly_over_q/properties.rs:18]:
```suggestion
    /// Returns true if the first coefficient is `1` and is the only coefficient.
```
Also for the other functions

File: src/rational/poly_over_q/properties.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    ///
```
Comment (MoogSven) [inline src/rational/poly_over_q/properties.rs:19]:
```suggestion
    ///
    /// # Examples
```

Comment (MoogSven) [reply src/rational/poly_over_q/properties.rs:19]: Also for the other functions

File: src/rational/poly_over_q/properties.rs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /// ensure that is_one returns `false` for other polynomials
+    #[test]
+    fn one_rejection() {
+        let small = PolyOverQ::from_str("4  1 0 0 1/123").unwrap();
+        let large = PolyOverQ::from_str(&format!("1  {}", (u128::MAX - 1) / 2 + 2)).unwrap();
+
+        assert!(!(small.is_one()));
+        assert!(!(large.is_one()));
```
Comment (MoogSven) [inline src/rational/poly_over_q/properties.rs:67]:
```suggestion
        assert!(!small.is_one());
        assert!(!large.is_one());
```
should be the same

File: src/rational/poly_over_q/properties.rs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /// ensure that is_zero returns `false` for non-zero polynomials
+    #[test]
+    fn zero_rejection() {
+        let small = PolyOverQ::from_str("4  0 0 0 1/8").unwrap();
+        let large = PolyOverQ::from_str(&format!("1  {}", (u128::MAX - 1) / 2 + 1)).unwrap();
+
+        assert!(!(small.is_zero()));
+        assert!(!(large.is_zero()));
```
Comment (MoogSven) [inline src/rational/poly_over_q/properties.rs:91]:
```suggestion
        assert!(!small.is_zero());
        assert!(!large.is_zero());
```

File: src/rational/mat_q/properties.rs (hunk 3 above)
Comment (MoogSven) [reply src/rational/mat_q/properties.rs:67]: Optionally also backticks around the function name.
```suggestion
    /// Ensure that `is_identity` returns `true` for identity matrices.
```

**Review by MoogSven — CHANGES_REQUESTED** [review]
Mostly comment changes:
- [x] Sentences should start with upper case and end with a dot. (Test case describtion, Returns [...]**.**)
- [x] Add `# Examples`

File: src/rational/mat_q/properties.rs (hunk 4 above)
Comment (MoogSven) [reply src/rational/mat_q/properties.rs:85]: Added in doc test

File: src/rational/poly_over_q/properties.rs (hunk 8 above)
Comment (MoogSven) [reply src/rational/poly_over_q/properties.rs:69]: Why did you not add this suggestion?

File: src/rational/q/properties.rs [hunk 10]
```diff
@@ -124,3 +154,51 @@ mod test_inv {
...
+    }
+
+    /// Ensure that is_zero returns `false` for non-zero values.
+    #[test]
+    fn zero_rejection() {
+        let small = Q::from(2);
+        let large = Q::from_str(&format!("{}", (u128::MAX - 1) / 2 + 1)).unwrap();
+
+        assert!(!(small.is_zero()));
+        assert!(!(large.is_zero()));
```
Comment (MoogSven) [inline src/rational/q/properties.rs:178]:
```suggestion
        assert!(!small.is_zero());
        assert!(!large.is_zero());
```

File: src/rational/mat_q/properties.rs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    use std::str::FromStr;
+
+    /// Ensure that is_square returns `true` for square matrices.
+    #[test]
+    fn square_detection() {
+        let mat_2x3 = MatQ::from_str("[[0, 4/9],[0, 0]]").unwrap();
+        let mat_3x2 = MatQ::from_str("[[0, 6/123, 4/7],[0, 0, 1/213],[4/341, 6/83, 1]]").unwrap();
+
+        assert!(mat_2x3.is_square());
+        assert!(mat_3x2.is_square());
```
Comment (MoogSven) [inline src/rational/mat_q/properties.rs:133]:
```suggestion
        let mat_2x2 = MatQ::from_str("[[0, 4/9],[0, 0]]").unwrap();
        let mat_3x3 = MatQ::from_str("[[0, 6/123, 4/7],[0, 0, 1/213],[4/341, 6/83, 1]]").unwrap();

        assert!(mat_2x2.is_square());
        assert!(mat_3x3.is_square());
```

**Review by MoogSven — APPROVED** [review]
Please check and fix the open conversation points before merging.

File: src/rational/mat_q/properties.rs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    /// Returns true if every diagonal entry of the upper square matrix is `1`
+    /// and all other entries are `0`.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::rational::MatQ;
+    /// use std::str::FromStr;
+    ///
+    /// let value = MatQ::from_str("[[1, 0],[0, 1]]").unwrap();
+    /// assert!(value.is_identity())
```
Comment (MoogSven) [inline src/rational/mat_q/properties.rs:26]: Add a semicolon at the end of your assert macros. This applies to all tests.

File: src/rational/poly_over_q/properties.rs (hunk 8 above)
Comment (3pmile) [reply src/rational/poly_over_q/properties.rs:69]: overlocked it

**Pull request merged**